pulsar/client/staging/up.py

```diff
@@ -388,6 +393,15 @@ def find_referenced_subfiles(self, directory):
         pattern = r'''[\'\"]?(%s%s[^\s\'\"]+)[\'\"]?''' % (escape(directory), escape(sep))
         return self.find_pattern_references(pattern)
 
+    def __ignore_tool_files(self, filename):
+        return filename in self.IGNORE_TOOL_FILES
+
+    def find_tool_files(self):
```

Comment:
If you can figure out how ``handle_transfer_source`` should actually work without that crazy hack then we could maybe do this by having a list of tool IDs that we do this for? The way we do with Galaxy lib exposure. No reason to break all of this and do all these unneeded transfers just for a handful of tools.

Reply:
Which part of it does "crazy hack" refer to? Sending everything for every tool?

i.e. is it ok to send everything for tools that we can determine need this treatment?

Reply:
> i.e. is it ok to send everything for tools that we can determine need this treatment?

Yes - this is fine - I always assumed we'd do this with file actions but if you want to do it this way I won't stand in your way.

The hack I'm referring to is the line in I said No to "pulsar/client/staging/up.py" - that just cannot be the right way to handle it. The inputs to that function aren't being setup properly or we're not handling nested files properly or something if that is the fix needed.

Reply:
Nevermind, I gather than "crazy hack" refers to your comment in `handle_transfer_source`.

Reply:
Yes - this is fine behavior if it is intentional.

Reply:
Whoops, deleted my comments after I figured out what you meant. 😂 

Thanks for clarifying.

Reply:
Could we perhaps detect extraneous files in the tool directory that aren't there in 95% of cases. i.e. If there is a helper script then we copy everything across. Otherwise nothing?

Reply:
What I mean is - If the current command line detection returns nothing then we copy nothing to pulsar, but if it returns anything or finds a `__tool_directory__` in the command line then it copies everything . I hope this makes sense.

Reply:
I *think* it should be pretty easy to limit this to tools that reference `$__tool_directory__` and tools that set `<command interpreter="">` by modifying `galaxy.tools:Tool` to store an attribute like `requires_tool_directory`, and then use that to determine which method should be used. Would that be reasonable to you, @jmchilton ?

Reply:
Scratch that, we're working on the list.

Reply:
It is a lot to ask but if we could turn the list into something like a list of file actions or regexes - that just default to ``**`` for the tools in the list I think that would be ideal.

- There will be tools that have big directories that we want to exclude stuff from and we can just make the list a dictionary.
- There will be tools that we need to capture the parent directory contents of - again we can just do that there also I think.
- Then we could expose tool XML constructs for overriding this list - which has been on the TODO list forever. We can take care of the long-term goals while getting these legacy things done. I'll try to help, I know this is a lot to ask and you're just trying to get shit working - sorry for being this way 😅.

Reply:
Can you let me know when you have a WIP Galaxy side of this?

Reply:
You mean this?

https://github.com/natefoo/galaxy/tree/unreferenced-tool-files